File: doc-examples/hibernate-reactive-example-java/src/main/java/example/BookRepository.java

package example;

import io.micronaut.data.annotation.Id;
import io.micronaut.data.annotation.Repository;
import io.micronaut.data.model.Page;
import io.micronaut.data.model.Pageable;
import io.micronaut.data.model.Slice;
import io.micronaut.data.repository.reactive.ReactorCrudRepository;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import javax.transaction.Transactional;
import java.util.function.Consumer;

// tag::repository[]
@Repository // <1>
interface BookRepository extends ReactorCrudRepository<Book, Long> { // <2>

    // tag::read[]
    Mono<Book> find(String title);

    Mono<BookDTO> findOne(String title);

    Flux<Book> findByPagesGreaterThan(int pageCount, Pageable pageable);

    Mono<Page<Book>> findByTitleLike(String title, Pageable pageable);

    Mono<Slice<Book>> list(Pageable pageable);
    // end::read[]

    @Transactional
    default Mono<Void> findByIdAndUpdate(Long id, Consumer<Book> bookConsumer) {
        return findById(id).map(book -> {
            bookConsumer.accept(book);
            return book;
        }).then();
    }

    // tag::save[]
    Mono<Book> save(Book entity);
    // end::save[]

    // tag::update[]
    Mono<Book> update(Book newBook);

    Mono<Void> update(@Id Long id, int pages);
    // end::delete[]

    // tag::delete[]
    @Override
    Mono<Long> deleteAll();

    Mono<Void> delete(String title);
    // end::delete[]
}
// end::repository[]

File: doc-examples/hibernate-reactive-example-python/src/main/python/example/BookRepository.py
from typing import Annotated

from jakarta.transaction import Transactional
from java.lang import Void
from java.util.function import Consumer
from micronaut.data.annotation import Id, Repository
from micronaut.data.model import Page, Pageable, Slice
from micronaut.data.repository.reactive import ReactorCrudRepository
from reactor.core.publisher import Flux, Mono

from example.Book import Book
from example.BookDTO import BookDTO


# tag::repository[]
@Repository  # <1>
class BookRepository(ReactorCrudRepository[Book, int]):  # <2>

    # tag::read[]
    def find(self, title: str) -> Mono[Book]: ...

    def findOne(self, title: str) -> Mono[BookDTO]: ...

    def findByPagesGreaterThan(self, pageCount: int, pageable: Pageable) -> Flux[Book]: ...

    def findByTitleLike(self, title: str, pageable: Pageable) -> Mono[Page[Book]]: ...

    def list(self, pageable: Pageable) -> Mono[Slice[Book]]: ...
    # end::read[]

    @Transactional
    def find_by_id_and_update(self, id: int, book_consumer: Consumer[Book]) -> Mono[Void]:
        def apply(book):
            book_consumer(book)
            return book
        return self.findById(id).map(apply).then()

    # tag::save[]
    def save(self, entity: Book) -> Mono[Book]: ...
    # end::save[]

    # tag::update[]
    def update(self, newBook: Book) -> Mono[Book]: ...

    def updatePagesById(self, id: Annotated[int, Id], pages: int) -> Mono[Void]: ...
    # end::update[]

    # tag::delete[]
    def deleteAll(self) -> Mono[int]: ...

    def delete(self, title: str) -> Mono[Void]: ...
    # end::delete[]
# end::repository[]
